### PR TITLE
feat: add problems tab in auxiliary panel & show warnings 

### LINKED ISSUE #941: Feature request: Discrepancies in behavior of graph builder and `engine` compiler
### Similar requests

- [X] I have searched and found no existing similar requests

### How are you using Studio?

_No response_

### What problems are you trying to solve?

We build the graph of meta-models, so does `engine`, but since we support edition, sometimes, we are stricter or less strict about several checks. This is a mega-thread to keep track of these cases:
- [ ] #660 
- [ ] #880

It's not wrong per se for our graph builder to fail, however, for users who migrated from `Pure IDE` to `Legend Studio` this could be a blocker since they don't have an environment to fix these errors productively.

> Note that we think these strict checks should still be present in Studio, we should eventually make these hard fails in even `engine` rather than _softening_ them because these are not something users `should or should not do`, not `must not do`, as such, we will keep this thread open till we completely pass migration phase and start to tighten graph builder checks again. 

### Describe the solution you would like to see implemented

The strategy we agreed upon is to relax our graph-builder and not to hard-fail on these errors, but only issue warnings, see https://github.com/finos/legend-studio/pull/883. `Engine` will now detect these issues and return them as warnings and we will block users from saving should we detect these kinds of problems in users' model

- [x] Change `engine` to return this as a warning as part of the compilation result - @pierredebelen 
- [x] ~~Disable saving when there are warnings~~ - @MauricioUyaguari and I think this is too disruptive at the moment and so we won't do it in the first go.
- [ ] `MAYBE` show a tab in `auxiliary panel` for warnings found. We can also show somewhere in the UI that we have warnings, for this to work, we need to show it in the status bar somewhere (see `vscode`) and consider doing something similar to what we have in https://github.com/finos/legend-studio/issues/1368 - e.g. every time we hit F9, store a hashcode
- [ ] Since there's no good way to reveal these warnings in the form, fall back to text mode, if we do the above step, clicking on the warnings will bring users to `text mode` and show the warning in yellow squiggle underlines
- [ ] We should have an application flag/config to enable throwing these errors in `graph-builder`: these errors are the ones mentioned above, this way, we could test these validations better. In production environment, we could turn this flag off so we don't throw error, otherwise, we will just print warning about this

### Describe alternatives you have considered

_No response_

### Documentation, Design, Adoption, Migration Strategy

_No response_

### Contribution

- [ ] I would like to work on this feature

### PR DESCRIPTION
**Pull request by xannem**
## Summary

fixes #941 

<h3> Form Mode </h3> 

- [x] show a tab in auxiliary panel for warnings found and in the status bar 

image

Kapture 2022-10-25 at 17 28 19

<hr/> 
<h3> Text Mode </h3> 

If you are in the text mode, any error or warning messages shown will also have the line location and will navigate to where the problem is located on click.
image

Kapture 2022-10-25 at 17 31 23

<hr/> 

<h3> Strict Flag </h3> 
There is now a toggle section in the developer tools panel where you can turn on the `strict` mode. This will mean that upon receiving errors, rather than passing and merely notifying users  the presence of warnings, Studio will throw an error and the graph itself will not be built. This is meant as an option to aid developers and is turned off unless manually toggled. 

Kapture 2022-10-25 at 19 56 19

## How did you test this change?

- [ ] Test(s) added
- [x] Manual testing (please provide screenshots/recordings)
- [ ] No testing (please provide an explanation)

### CODE REVIEW

**Review by akphi — CHANGES_REQUESTED** [review]

File: packages/legend-art/style/base/_panel.scss [hunk 1]
```diff
@@ -252,6 +256,18 @@
 
   &__section {
     padding: 0 1rem;
+
+    .input-group__error-message {
+      max-width: 45rem;
+    }
+  }
+
+  &__icon {
```
[the remark this answers is not in the document]
Comment (xannem) [reply packages/legend-art/style/base/_panel.scss:265]: only `PanelFormTextEditor` inputs have the qualification that their max-width is 45rem so the dropdown of an invalid error i put the edit in here as well bc otherwise the red space will expand to the entirety of the default 100% width 🤔

File: packages/legend-graph/src/graphManager/action/EngineWarning.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ApplicationError } from '@finos/legend-shared';
+import type { SourceInformation } from './SourceInformation.js';
+
+export class EngineWarning extends ApplicationError {
```
[the remark this answers is not in the document]
Comment (xannem) [reply packages/legend-graph/src/graphManager/action/EngineWarning.ts:20]: hm not really i originally had it as an own class but was thinking maybe it should count as an error technically with the future plans. but will switch to not extend since true it is called `EngineWarning` 👍

File: packages/legend-graph/src/graphManager/action/EngineWarning.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  message: string | undefined;
+
+  constructor(message: string, sourceInformation: SourceInformation) {
+    this.message = message;
+
+    this.sourceInformation = sourceInformation;
+  }
+}
+
+export class EntitiesAndWarnings {
```
Comment (akphi) [inline packages/legend-graph/src/graphManager/action/EngineWarning.ts:31]: Let's move this to `graphManager/action/compilation/CompilationResult.ts`

File: packages/legend-graph/src/graphManager/AbstractPureGraphManager.ts [hunk 4]
```diff
@@ -238,13 +242,17 @@ export abstract class AbstractPureGraphManager {
...
+      getCompilationWarnings?: boolean;
+    },
+  ): Promise<EngineWarning[] | undefined>;
+  abstract compileTextWithEntitiesAndWarnings(
     graphGrammar: string,
     graph: PureModel,
-    options?: { onError?: () => void },
-  ): Promise<Entity[]>;
+    options?: { onError?: () => void; getCompilationWarnings?: boolean },
+  ): Promise<EntitiesAndWarnings>;
```
Comment (akphi) [inline packages/legend-graph/src/graphManager/AbstractPureGraphManager.ts:255]: --> `Promise<{ warnings?: EngineWarning[] | undefined, entities: Entity[] }>`

File: packages/legend-graph/src/graphManager/protocol/pure/v1/engine/V1_Engine.ts [hunk 5]
```diff
@@ -387,11 +411,46 @@ export class V1_Engine {
...
+          V1_CompilationError.serialization.fromJson(
+            error.payload as PlainObject<V1_CompilationError>,
+          ),
+        );
+      }
+      throw error;
+    }
+  }
+
+  async compileTextWithEntitiesAndWarnings(
```
Comment (akphi) [inline packages/legend-graph/src/graphManager/protocol/pure/v1/engine/V1_Engine.ts:449]: --> `compileText`

File: packages/legend-graph/src/graphManager/protocol/pure/v1/engine/V1_Engine.ts [hunk 6]
```diff
@@ -407,7 +466,18 @@ export class V1_Engine {
       : mainGraph;
     try {
       await this.engineServerClient.compile(pureModelContextDataJson);
-      return V1_deserializePureModelContextData(mainGraph);
+      const entities = V1_deserializePureModelContextData(mainGraph);
+      const compileResponse = await this.engineServerClient.compile(
```
Comment (akphi) [inline packages/legend-graph/src/graphManager/protocol/pure/v1/engine/V1_Engine.ts:470]: --> `compilationResult`

File: packages/legend-graph/src/graphManager/protocol/pure/v1/engine/V1_Engine.ts [hunk 7]
```diff
@@ -387,11 +411,46 @@ export class V1_Engine {
...
 
-  async compileText(
+  async getWarningsFromCompilePureModelContextData(
+    model: V1_PureModelContextData,
+    options?:
+      | { onError?: (() => void) | undefined; getCompilationWarnings?: boolean }
+      | undefined,
+  ): Promise<EngineWarning[] | undefined> {
+    try {
+      const compileResponse = await this.engineServerClient.compile(
```
Comment (akphi) [inline packages/legend-graph/src/graphManager/protocol/pure/v1/engine/V1_Engine.ts:421]: --> `compilationResult`

File: packages/legend-graph/src/graphManager/protocol/pure/v1/engine/V1_Engine.ts [hunk 8]
```diff
@@ -387,11 +411,46 @@ export class V1_Engine {
     }
   }
 
-  async compileText(
+  async getWarningsFromCompilePureModelContextData(
```
Comment (akphi) [inline packages/legend-graph/src/graphManager/protocol/pure/v1/engine/V1_Engine.ts:414]: Let's remove this function, don't think this gets used anymore

File: packages/legend-graph/src/graphManager/protocol/pure/v1/engine/V1_Engine.ts [hunk 9]
```diff
@@ -362,14 +376,24 @@ export class V1_Engine {
...
     try {
-      await this.engineServerClient.compile(
+      const compileResponse = await this.engineServerClient.compile(
         this.serializePureModelContextData(model),
       );
+      if (
+        compileResponse.warnings &&
+        options?.getCompilationWarnings === true
+      ) {
+        return compileResponse.warnings as EngineWarning[];
```
Comment (akphi) [inline packages/legend-graph/src/graphManager/protocol/pure/v1/engine/V1_Engine.ts:393]: This is not proper, you need to create a shape for `V1_CompilationResult` with proper serialization and a corresponding `CompilationResult`

File: packages/legend-graph/src/graphManager/action/EngineWarning.ts (hunk 3 above)
Comment (akphi) [reply packages/legend-graph/src/graphManager/action/EngineWarning.ts:31]: This should be renamed to `TextCompilationResult`

File: packages/legend-graph/src/graphManager/action/EngineWarning.ts [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { Entity } from '@finos/legend-storage';
+import type { SourceInformation } from './SourceInformation.js';
+
+export class EngineWarning {
```
Comment (akphi) [inline packages/legend-graph/src/graphManager/action/EngineWarning.ts:20]: Let's rename this and the file to `CompilationWarning`. Let's move this whole file to `graphManager/action/compilation/CompilationResult.ts`

File: packages/legend-art/src/utils/TextEditorUtils.ts [hunk 11]
```diff
@@ -118,6 +120,17 @@ export const setErrorMarkers = (
   ]);
 };
 
+export const setAllWarningMarkers = (
+  editorModel: monacoEditorAPI.ITextModel,
+  allWarnings: monacoEditorAPI.IMarkerData[],
+): void => {
+  monacoEditorAPI.setModelMarkers(editorModel, DUMMY_MARKER_OWNER, allWarnings);
```
Comment (akphi) [inline packages/legend-art/src/utils/TextEditorUtils.ts:127]: --> `setWarningMarkers()`. not sure why we parse in `IMarkerData`, we should follow `setErrorMarkers`, we enforce the severity and only let people pass in a list of coordinates (i.e. `TextEditorRange`)

Let's create 2 types:

```ts
export type TextEditorPosition = {
  lineNumber: number;
  column: number;
};

export type TextEditorRange = {
  startLineNumber: number;
  startColumn: number;
  endLineNumber: number;
  endColumn: number;
};

```

File: packages/legend-art/src/components/panel/Panel.tsx [hunk 12]
```diff
@@ -66,15 +71,17 @@ export const PanelHeaderActionItem: React.FC<{
   className?: string;
   disabled?: boolean;
   children: React.ReactNode;
+  tabIndex?: number;
```
Comment (akphi) [inline packages/legend-art/src/components/panel/Panel.tsx:74]: no need `tabIndex` I think, we can almost also default to `-1`

File: packages/legend-application-studio/src/components/editor/aux-panel/AuxiliaryPanel.tsx [hunk 13]
```diff
@@ -97,12 +114,18 @@ export const AuxiliaryPanel = observer(() => {
                 )}
                 <div className="auxiliary-panel__header__tab__title">
                   {tab.name}
+                  {(tab.counter || tab.counter === 0) && (
```
Comment (akphi) [inline packages/legend-application-studio/src/components/editor/aux-panel/AuxiliaryPanel.tsx:117]: --> `tab.couner !== undefined`

File: packages/legend-application-studio/src/components/editor/aux-panel/Problems.tsx [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+      editorStore.graphState.mostRecentFormModeCompilationGraphHash !==
+      editorStore.graphState.currentGraphHash;
+  }
+
+  const setWarning = (warning: EngineWarning): void => {
+    editorStore.grammarTextEditorState.setLocationForceCursor(warning);
+  };
+
+  return (
+    <Panel className="console-panel">
```
Comment (akphi) [inline packages/legend-application-studio/src/components/editor/aux-panel/Problems.tsx:90]: don't reuse CSS classes like this

File: packages/legend-application-studio/src/stores/ChangeDetectionState.ts [hunk 15]
```diff
@@ -271,12 +272,14 @@ export class ChangeDetectionState {
   aggregatedConflictResolutionChanges: EntityDiff[] = [];
   conflicts: EntityChangeConflict[] = []; // conflicts in conflict resolution mode (derived from aggregated workspace changes and conflict resolution changes)
   resolutions: EntityChangeConflictResolution[] = [];
+  graphHash: string | undefined;
```
Comment (akphi) [inline packages/legend-application-studio/src/stores/ChangeDetectionState.ts:275]: come to think about it, don't think we need to store this here

File: packages/legend-application-studio/src/stores/ChangeDetectionState.ts [hunk 16]
```diff
@@ -335,12 +339,29 @@ export class ChangeDetectionState {
...
+    this.graphHash = hashArray(
+      this.editorStore.graphManagerState.graph.allOwnElements,
+    );
   }
 
   setAggregatedProjectLatestChanges(diffs: EntityDiff[]): void {
     this.aggregatedProjectLatestChanges = diffs;
   }
 
+  snapshotGraphHash(): void {
```
Comment (akphi) [inline packages/legend-application-studio/src/stores/ChangeDetectionState.ts:352]: therefore, we don't need this method, whoever need to get the current graph hash, just call `getCurrentGraphHash`

File: packages/legend-application-studio/src/stores/project-viewer/ProjectViewerStore.ts [hunk 17]
```diff
@@ -476,7 +476,6 @@ export class ProjectViewerStore {
         yield flowResult(
           this.editorStore.graphState.globalCompileInTextMode({
             ignoreBlocking: true,
-            suppressCompilationFailureMessage: true,
```
Comment (akphi) [inline packages/legend-application-studio/src/stores/project-viewer/ProjectViewerStore.ts:479]: why?

File: packages/legend-application-studio/src/stores/editor-state/GrammarTextEditorState.ts [hunk 18]
```diff
@@ -63,12 +64,13 @@ export class GrammarTextEditorState {
   graphGrammarText = '';
   currentElementLabelRegexString?: string | undefined;
   wrapText = false;
-  error?: EngineError | undefined;
+  locationForceCursor: EngineWarning | EngineError | undefined;
```
Comment (akphi) [inline packages/legend-application-studio/src/stores/editor-state/GrammarTextEditorState.ts:67]: let's do `forcedCursorPosistion?: TextEditorPosition | undefined`

File: packages/legend-application-studio/src/stores/editor-state/GrammarTextEditorState.ts [hunk 19]
```diff
@@ -78,8 +80,10 @@ export class GrammarTextEditorState {
     this.editorStore = editorStore;
   }
 
-  setError(error: EngineError | undefined): void {
-    this.error = error;
+  setLocationForceCursor(
```
Comment (akphi) [inline packages/legend-application-studio/src/stores/editor-state/GrammarTextEditorState.ts:83]: --> `forceCursorPosition(position: TextEditorPosition | undefined)`

File: packages/legend-application-studio/src/components/editor/edit-panel/GrammarTextEditor.tsx [hunk 20]
```diff
@@ -811,10 +812,14 @@ export const GrammarTextEditor = observer(() => {
...
         grammarTextEditorState.setGraphGrammarText(getEditorValue(_editor));
-        editorStore.graphState.clearCompilationError();
         // we can technically can reset the current element label regex string here
         // but if we do that on first load, the cursor will not jump to the current element
         // also, it's better to place that logic in an effect that watches for the regex string
+
+        editorStore.graphState.clearCompilationError();
+
+        clearProblemMarkers();
+        editorStore.graphState.setIsStaleTextWarnings(true);
```
Comment (akphi) [inline packages/legend-application-studio/src/components/editor/edit-panel/GrammarTextEditor.tsx:822]: I think we talked about this, this should actually just clear the warnings, there's not even need to call `setIsStale...` at that point

File: packages/legend-application-studio/src/components/editor/edit-panel/GrammarTextEditor.tsx [hunk 21]
```diff
@@ -882,17 +887,32 @@ export const GrammarTextEditor = observer(() => {
     const editorModel = editor.getModel();
     if (editorModel) {
       editorModel.updateOptions({ tabSize: TAB_SIZE });
-      if (error?.sourceInformation) {
-        setErrorMarkers(
+      const collectionOfWarnings: {
```
Comment (akphi) [inline packages/legend-application-studio/src/components/editor/edit-panel/GrammarTextEditor.tsx:890]: this looks like we no longer alert about the error, why can't we separately call `setErrorMarkers` and `setWarningMarkers`?

File: packages/legend-application-studio/src/stores/EditorGraphState.ts [hunk 22]
```diff
@@ -118,6 +126,9 @@ export interface GraphBuilderResult {
 export class EditorGraphState {
   editorStore: EditorStore;
   graphGenerationState: GraphGenerationState;
+  warnings?: EngineWarning[] | undefined;
+  mostRecentFormModeCompilationGraphHash: string | undefined;
+  isStaleTextWarnings: boolean;
```
Comment (akphi) [inline packages/legend-application-studio/src/stores/EditorGraphState.ts:131]: we can remove this now

File: packages/legend-application-studio/src/stores/editor-state/GrammarTextEditorState.ts [hunk 23]
```diff
@@ -63,12 +64,13 @@ export class GrammarTextEditorState {
   graphGrammarText = '';
   currentElementLabelRegexString?: string | undefined;
   wrapText = false;
-  error?: EngineError | undefined;
```
Comment (akphi) [inline packages/legend-application-studio/src/stores/editor-state/GrammarTextEditorState.ts:66]: where did you move this error to? `EditorGraphState`?

File: packages/legend-application-studio/src/stores/EditorGraphState.ts [hunk 24]
```diff
@@ -127,28 +138,59 @@ export class EditorGraphState {
...
       graphGenerationState: false,
+      warnings: true,
       getPackageableElementType: false,
-      hasCompilationError: computed,
       clearCompilationError: action,
+      setWarnings: action,
+      setIsStaleTextWarnings: action,
+      setMostRecentFormModeCompilationGraphHash: action,
+      hasCompilationError: computed,
+      currentGraphHash: computed,
```
Comment (akphi) [inline packages/legend-application-studio/src/stores/EditorGraphState.ts:152]: don't need this, just call `return this.editorStore.changeDetectionState.graphHash;` directly

File: packages/legend-application-studio/src/stores/EditorGraphState.ts [hunk 25]
```diff
@@ -127,28 +138,59 @@ export class EditorGraphState {
...
     this.graphGenerationState = new GraphGenerationState(this.editorStore);
+    this.warnings = undefined;
+    this.isStaleTextWarnings = false;
+  }
+
+  setWarnings(val: EngineWarning[] | undefined): void {
+    this.warnings = val;
+  }
+
+  setIsStaleTextWarnings(val: boolean): void {
```
Comment (akphi) [inline packages/legend-application-studio/src/stores/EditorGraphState.ts:165]: let's remove this

File: packages/legend-application-studio/src/stores/EditorGraphState.ts [hunk 26]
```diff
@@ -127,28 +138,59 @@ export class EditorGraphState {
...
+    this.isStaleTextWarnings = val;
+  }
+
+  setMostRecentFormModeCompilationGraphHash(
+    graphHash: string | undefined,
+  ): void {
+    this.mostRecentFormModeCompilationGraphHash = graphHash;
```
Comment (akphi) [inline packages/legend-application-studio/src/stores/EditorGraphState.ts:172]: don't need this setter, we can just assign the value directly in the body of the `compile...()` generators

File: packages/legend-application-studio/src/stores/EditorGraphState.ts [hunk 27]
```diff
@@ -594,33 +662,45 @@ export class EditorGraphState {
...
+        );
+      } else {
+        if (!options?.disableNotificationOnSuccess) {
+          this.editorStore.applicationStore.notifySuccess(
+            'Compiled successfully',
+          );
+        }
       }
 
+      this.editorStore.graphState.setWarnings(
```
Comment (akphi) [inline packages/legend-application-studio/src/stores/EditorGraphState.ts:692]: I don't see a need to do this type casting

File: packages/legend-application-studio/src/stores/EditorGraphState.ts [hunk 28]
```diff
@@ -461,6 +502,14 @@ export class EditorGraphState {
     disableNotificationOnSuccess?: boolean;
     openConsole?: boolean;
   }): GeneratorFn<FormModeCompilationOutcome> {
+    if (!this.mostRecentFormModeCompilationGraphHash) {
```
Comment (akphi) [inline packages/legend-application-studio/src/stores/EditorGraphState.ts:505]: why this complexity? why can't we just do 

```ts
this.mostRecentFormModeCompilationGraphHash = ...
```

File: packages/legend-graph/src/graphManager/AbstractPureGraphManager.ts [hunk 29]
```diff
@@ -238,13 +242,17 @@ export abstract class AbstractPureGraphManager {
   abstract compileEntities(entities: Entity[]): Promise<void>;
   abstract compileGraph(
     graph: PureModel,
-    options?: { onError?: () => void; keepSourceInformation?: boolean },
-  ): Promise<void>;
-  abstract compileText(
+    options?: {
+      onError?: () => void;
+      keepSourceInformation?: boolean;
+      getCompilationWarnings?: boolean;
```
Comment (akphi) [inline packages/legend-graph/src/graphManager/AbstractPureGraphManager.ts:248]: we don't need this flag, let's always return `warnings` if there is

File: packages/legend-graph/src/graphManager/AbstractPureGraphManager.ts [hunk 30]
```diff
@@ -238,13 +242,17 @@ export abstract class AbstractPureGraphManager {
...
+    },
+  ): Promise<EngineWarning[] | undefined>;
+  abstract compileTextWithEntitiesAndWarnings(
     graphGrammar: string,
     graph: PureModel,
-    options?: { onError?: () => void },
-  ): Promise<Entity[]>;
+    options?: { onError?: () => void; getCompilationWarnings?: boolean },
```
Comment (akphi) [inline packages/legend-graph/src/graphManager/AbstractPureGraphManager.ts:254]: let's remove the warnings flag

File: packages/legend-application-studio/src/stores/EditorGraphState.ts [hunk 31]
```diff
@@ -594,33 +662,45 @@ export class EditorGraphState {
...
+      this.editorStore.graphState.setWarnings(
+        compilationWarnings as EngineWarning[],
+      );
+
       yield flowResult(this.updateGraphAndApplication(entities));
     } catch (error) {
       assertErrorThrown(error);
       if (error instanceof EngineError) {
-        this.editorStore.grammarTextEditorState.setError(error);
+        this.editorStore.grammarTextEditorState.setLocationForceCursor(error);
```
Comment (akphi) [inline packages/legend-application-studio/src/stores/EditorGraphState.ts:700]: let's revert the change you make to `error` in `GrammarTextEditorState` though, we still want to keep the error around. We can't coalesce it with warnings like what we're doing now.
